### PR TITLE
Add support for 4byteTracer in debug_trace* methods

### PR DESCRIPTION
**Pull request by usmansaleem**
## PR description
Implements Geth's `4byteTracer` for `debug_traceTransaction`, `debug_traceBlock`, and related debug trace methods. The 4byteTracer collects function selectors (the first 4 bytes of call data) from all internal calls made during transaction execution, along with the size of the remaining call data (excluding the 4-byte selector).

### Changes

  - Add `FOUR_BYTE_TRACER` to `TracerType` enum
  - Create `FourByteTracerResult` class for JSON output formatting
  - Create `FourByteTracerResultConverter` to process transaction traces
  - Wire 4ByteTracer into `DebugTraceTransactionStepFactory`
  - Add comprehensive unit tests 

### Implementation Details
The implementation matches Geth's 4byteTracer behavior:
  - Only processes CALL-type operations (CALL, CALLCODE, DELEGATECALL, STATICCALL)
  - Excludes CREATE/CREATE2 operations and precompiled contracts
  - Only counts calls that successfully enter (depth increases in next frame)
  - Calculates size as `input_length - 4` (excluding the selector)
  - Correctly handles nested sub-calls at any depth with proper occurrence counting
  - Uses ProtocolSpec's precompile registry for accurate precompile detection across different protocol specifications
  - Injects ProtocolSpec into DebugTraceTransactionStepFactory for protocol-aware tracing

### Output Format

  Returns a map where keys are in the format `"0x[4-byte-selector]-[remaining-size]"` and values are occurrence counts:
  ```json
  {
    "0x27dc297e-128": 1,
    "0x38cc4831-0": 2,
    "0x524f3889-96": 1
  }
```

###  Testing

  - Added FourByteTracerResultConverterTest with comprehensive test cases covering:
    - Null/empty input handling
    - Function selector extraction from CALL operations
    - Filtering (precompiles, non-entered calls, CREATE ops, short input)
    - Multiple calls with same/different selectors
    - All CALL-type operations (CALL, CALLCODE, DELEGATECALL, STATICCALL)
    - Nested sub-calls with multiple occurrences
    - Edge cases (exactly 4 bytes of input)
  - Updated DebugTraceTransactionStepFactoryTest to include 4ByteTracer
  - All existing tests pass

###  References

  - https://geth.ethereum.org/docs/developers/evm-tracing/built-in-tracers#4byte-tracer

### Acknowledgements
Based on original contribution PR: https://github.com/hyperledger/besu/pull/9202 by [@JukLee0ira](https://github.com/JukLee0ira)

### Thanks for sending a pull request! Have you done the following?

- [x] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [x] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [x] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [x] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [x] spotless: `./gradlew spotlessApply`
- [x] unit tests: `./gradlew build`
- [x] acceptance tests: `./gradlew acceptanceTest`
- [x] integration tests: `./gradlew integrationTest`
- [x] reference tests: `./gradlew ethereum:referenceTests:referenceTests`
- [x] hive tests: [Engine or other RPCs modified?](https://lf-hyperledger.atlassian.net/wiki/spaces/BESU/pages/22156302/Using+Hive+Test+Suite)

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverter.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+      return;
+    }
+
+    // Skip precompiled contracts (addresses 0x01-0x0a typically)
+    // Note: This logic should match how precompiles are detected elsewhere in Besu
+    // For now, simple check for addresses <= 0x0a
+    final Bytes toAddressBytes = to.get().getBytes();
+    if (toAddressBytes.numberOfLeadingZeroBytes() >= 19
+        && toAddressBytes.trimLeadingZeros().size() == 1
+        && toAddressBytes.trimLeadingZeros().get(0) <= 0x0a) {
```
Comment (Matilda-Clerke) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverter.java]: Would be good use constants for magic numbers 19 and 0x0a

Comment (usmansaleem) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverter.java]: This will be removed once https://github.com/hyperledger/besu/pull/9668/ is approved and merge in.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverter.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+ *
+ * <p>For each qualifying call operation, it extracts the first 4 bytes of the input data (the
+ * function selector) and combines it with the size of the remaining call data (size - 4) to create
+ * a key in the format "0x[4-byte-selector]-[remaining-size]". The result is a map of these keys to
+ * their occurrence counts.
+ *
+ * @see <a href="https://github.com/ethereum/go-ethereum/blob/master/eth/tracers/native/4byte.go">
+ *     Geth 4byteTracer Implementation</a>
+ */
+public class FourByteTracerResultConverter {
```
Comment (lu-pinto) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverter.java]: make class final too

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverterTest.java [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+  }
+
+  @Test
+  @DisplayName("should throw NullPointerException when protocolSpec is null")
+  void shouldThrowNullPointerExceptionWhenProtocolSpecIsNull() {
+    TransactionTrace mockTrace = createMockTraceWithContractCreation();
+    assertThatThrownBy(() -> FourByteTracerResultConverter.convert(mockTrace, null))
+        .isInstanceOf(NullPointerException.class)
+        .hasMessageContaining("FourByteTracerResultConverter requires a non-null ProtocolSpec");
+  }
```
Comment (lu-pinto) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverterTest.java]: I think these assertions for NPEs are overkill. NPEs are programmer errors not expected behaviour

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverterTest.java [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+    // Then
+    assertThat(result).isNotNull();
+    Map<String, Integer> counts = result.getSelectorCounts();
+    assertThat(counts).hasSize(1);
+    assertThat(counts).containsEntry("0x12345678-0", 1);
+  }
+
+  @Test
+  @DisplayName("should handle nested sub-calls and count multiple occurrences")
+  void shouldHandleNestedSubCallsAndCountMultipleOccurrences() {
```
Comment (lu-pinto) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverterTest.java]: tests named with ...And... generally mean you need 2 separate tests. This test is more than 100 lines long!
I'm even for duplicating setup code that asserts different things, this way unit tests are cleaner.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverterTest.java (hunk 3 above)
Comment (usmansaleem) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverterTest.java]: agreed, removed.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverter.java (hunk 2 above)
Comment (usmansaleem) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverter.java]: done.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverterTest.java (hunk 4 above)
Comment (usmansaleem) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/FourByteTracerResultConverterTest.java]: 👍🏼 Refactored it into two methods.

**Comment by usmansaleem** [comment]
@lu-pinto I've also added spec tests generated against Geth for 4ByteTracer.
